py_retro/bsv_input.py

```diff
@@ -82,20 +79,25 @@ def set_input_state_file(core, filename, restore=True, expectedCartCRC=None):
     filename to use, rather than a function.
     """
 
-    generator = bsv_decode(filename)
+    bsv = BSV(filename)
 
-    def wrapper(port, device, index, id):
-        return next(generator)
-
-    (serializerVersion, cartCRC, saveStateData) = next(generator)
-
-    if expectedCartCRC is not None:
+    if expectedCartCRC is not None and bsv.cart_crc != expectedCartCRC:
         raise CartMismatch("Movie is for cart with CRC32 %r, expected %r"
-                % (cartCRC, expectedCartCRC))
+                % (bsv.cart_crc, expectedCartCRC))
 
     if restore:
-        core.unserialize(saveStateData)
-
-    core.set_input_state_cb(wrapper)
-    return wrapper
+        # retry loop for the multi-threaded ParaLLEl-N64,
+        # which refuses to load state while 'initializing'
+        for i in range(100):
+            # noinspection PyBroadException
+            try:
+                core.unserialize(bsv.state_data)
```

Comment:
the core wrapper throws when the core returns `false` for the call to `bool retro_unserialize(...)` (if the C code crashes, the python interp goes down with it, so no catch there)

parallel-n64 is threaded, and its unserialize returns false when the secondary thread is still "initializing."  so we hack around it by letting it fail a bunch of times.

it's worth mentioning that we re-raise the exception at the end of the loop anyway, just in case it *is* worth seeing :)